### PR TITLE
feat: Copy on write optimization for brillig

### PR DESCRIPTION
**Pull request by sirasistant**
# Description

## Problem\*

Resolves 

## Additional Context

This PR supersedes https://github.com/noir-lang/noir/pull/3118.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: Cargo.lock
Comment (TomAFrench) [inline Cargo.lock:1]: I think we can revert these changes as they seem unrelated to the rest of the PR.

**Review by jfecher — APPROVED** [review]
Since https://github.com/noir-lang/noir/pull/3593 is merged, this is now good to go
